rfcs/0161-intelligent-worker-cycling.md

```diff
@@ -0,0 +1,112 @@
+  configurations (`workermanager.registerWorker` already includes `workerConfig`)
+* Workers (via Worker Runner) will refresh their configs from successful
+  `workermanager.reregisterWorker` calls, or gracefully shutdown when receiving
+  a HTTP 410 status code from either `workermanager.registerWorker` or
+  `workermanager.reregisterWorker`
+  
+
+## Worker Manager changes
+
+Depending on the provider, Worker Pools contain different data. Regardless, the
+logic to decide if a worker pool launch configuration has changed is the same:
+
+* If Worker Manager can determine that the parameters used to launch an
+  existing running worker (instance type, disk sizes, region, ...) from a
+  previous version of the worker pool definition are still a valid combination
+  of parameters against the latest worker pool configuration, responses to
+  `workermanager.registerWorker` or `workermanager.reregisterWorker` will be
+  treated as before, and the worker will not be earmarked for decommission.
+* Otherwise, if Worker Manager is either unable to determine that the launch
+  parameters of a given worker are still valid, or is able to determine that
+  they are no longer consistent with the latest worker pool definition, it should:
+
+    * Respond to `workermanager.registerWorker` and `workermanager.reregisterWorker`
+      calls with an HTTP 410 response status code
+    * Terminate the worker (in the case of non-static workers) as soon as
+      possible, but _no earlier_ than 15 minutes after the worker's current
+      credentials have expired
+    * Quarantine the worker (in case of static workers) as soon as possible,
+      but _no earlier_ than 15 minutes after the worker's current credentials
+      have expired. Note, currently this will have no effect, since static worker
+      pools do not contain environment state, but this will allow us to include
+      environment state in static worker pool definitions in future, which can be
+      compared to data provided by the worker in `workermanager.registerWorker`
+      calls. This way, static workers with outdated state may be automatically
+      decommissioned. Note, there is currently no means to provide a quarantine
+      reason in `queue.quarantineWorker`, but if that becomes available, in future
+      it should be used to record the reason. 
+
+* Worker Manager should return the _most recent_ worker pool `workerConfig` for
+  a worker in the `reregisterWorker` response, like it already does for
+  `registerWorker`.
+
+## Worker changes
+
+The only supported way to run taskcluster-developed workers under Worker
+Manager is via Worker Runner. However, if you are running your own worker
+implementations under Worker Manager that do not work with Worker Runner
+(scriptworker?) then either the following changes will also be needed in your
+worker implementations, or you will need to adapt your worker implementations
+to run under Worker Runner.
```

Comment:
We're planning to de-support standalone workers, current proposal being to do so in January 2021.  So I don't think we need to worry about this scenario too much.